packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,216 @@
+	revertibles: IntervalRevertible[],
+) {
+	const startSeg = interval.start.getSegment() as SharedStringSegment;
+	const endSeg = interval.end.getSegment() as SharedStringSegment;
+	const startRef = string.createLocalReferencePosition(
+		startSeg,
+		interval.start.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeBegin,
+		undefined,
+	);
+	const endRef = string.createLocalReferencePosition(
+		endSeg,
+		interval.end.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeEnd,
+		undefined,
+	);
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: startRef,
+		end: endRef,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalChangeToRevertibles(
+	string: SharedString,
+	newInterval: SequenceInterval,
+	previousInterval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+) {
+	const startSeg = previousInterval.start.getSegment() as SharedStringSegment;
+	const endSeg = previousInterval.end.getSegment() as SharedStringSegment;
+	const prevStartRef = string.createLocalReferencePosition(
+		startSeg,
+		previousInterval.start.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeBegin,
+		undefined,
+	);
+	const prevEndRef = string.createLocalReferencePosition(
+		endSeg,
+		previousInterval.end.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeEnd,
+		undefined,
+	);
+	revertibles.push({
+		event: IntervalEventType.CHANGE,
+		interval: newInterval,
+		start: prevStartRef,
+		end: prevEndRef,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalPropertyChangedToRevertibles(
+	interval: SequenceInterval,
+	deltas: PropertySet,
+	revertibles: IntervalRevertible[],
+) {
+	revertibles.push({
+		event: IntervalEventType.PROPERTYCHANGED,
+		interval,
+		propertyDeltas: deltas,
+	});
+
+	return revertibles;
+}
+// Uses of referenceRangeLabels will be removed once AB#4081 is completed.
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	const id = getId(revertible.interval);
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	string.getIntervalCollection(label).removeIntervalById(id);
+}
+
+function revertLocalDelete(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.DELETE>,
+) {
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	const type = revertible.interval.intervalType;
+	// reusing the id causes eventual consistency bugs, so it is removed here and recreated in add
+	const { intervalId, ...props } = revertible.interval.properties;
+	const int = string.getIntervalCollection(label).add(start, end, type, props);
+
+	idMap.forEach((oldId, newId) => {
+		if (intervalId === newId) {
+			idMap.set(oldId, getId(int));
+		}
+	});
+	idMap.set(intervalId, getId(int));
```

Comment:
Whoops, sorry. I told you to call getId in this function, but I meant for re-creating the interval. However, that actually makes no sense. Change it back to what you had before since it will never be in the map.

```suggestion
	idMap.set(intervalId, int.getIntervalId());
```